agent/taskresource/credentialspec/credentialspec.go

```diff
@@ -0,0 +1,556 @@
+	"github.com/aws/amazon-ecs-agent/agent/s3"
+	s3factory "github.com/aws/amazon-ecs-agent/agent/s3/factory"
+	"github.com/aws/amazon-ecs-agent/agent/ssm"
+	ssmfactory "github.com/aws/amazon-ecs-agent/agent/ssm/factory"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	resourcestatus "github.com/aws/amazon-ecs-agent/agent/taskresource/status"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+	"github.com/aws/amazon-ecs-agent/agent/utils/oswrapper"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+const (
+	// ResourceName is the name of the credentialspec resource
+	ResourceName = "credentialspec"
+
+	tempFileName = "temp_file"
+
+	// filePerm is the permission for the credentialspec file.
+	filePerm = 0755
+
+	s3DownloadTimeout = 30 * time.Second
+)
+
+var (
+	// resourceDir will be populated by environment variables
+	resourceDir = ""
+)
+
+func init() {
+	// TODO: Use registry
+	// This should always be available on Windows instances
+	appDataDir := os.Getenv("APPDATA")
+	if appDataDir != "" {
+		resourceDir = appDataDir
+	} else {
+		tempDir := os.Getenv("TEMP")
+		if tempDir != "" {
+			resourceDir = tempDir
+		}
+	}
+
+	if resourceDir == "" {
+		seelog.Critical("Unable to obtain valid credentialspec resource dir")
+		os.Exit(1)
+	}
+}
+
+// CredentialSpecResource is the abstraction for credential spec resources
+type CredentialSpecResource struct {
+	taskARN string
+	region  string
+
+	credentialsManager     credentials.Manager
+	executionCredentialsID string
+
+	ioutil ioutilwrapper.IOUtil
+	os     oswrapper.OS
+
+	createdAt           time.Time
+	desiredStatusUnsafe resourcestatus.ResourceStatus
+	knownStatusUnsafe   resourcestatus.ResourceStatus
+	// appliedStatus is the status that has been "applied" (e.g., we've called some
+	// operation such as 'Create' on the resource) but we don't yet know that the
+	// application was successful, which may then change the known status. This is
+	// used while progressing resource states in progressTask() of task manager
+	appliedStatus                      resourcestatus.ResourceStatus
+	resourceStatusToTransitionFunction map[resourcestatus.ResourceStatus]func() error
+
+	// terminalReason should be set for resource creation failures. This ensures
+	// the resource object carries some context for why provisioning failed.
+	terminalReason     string
+	terminalReasonOnce sync.Once
+
+	// ssmClientCreator is a factory interface that creates new SSM clients. This is
+	// needed mostly for testing.
+	ssmClientCreator ssmfactory.SSMClientCreator
+
+	// s3ClientCreator is a factory interface that creates new S3 clients. This is
+	// needed mostly for testing.
+	s3ClientCreator s3factory.S3ClientCreator
+
+	// required for processing credential specs, key is input credential spec
+	requiredCredentialSpecs map[string][]*apicontainer.Container
+	// map to transform credential spec values, key is a input credential spec
```

Comment:
The description comment for **credSpecMap** is not clear enough, can you add an example here?